[Linux] TextureMapperPlatformLayerProxyDMABuf should support different colorspaces

zdobersek opened the pull request:
#### 733ff10e75f6da2019decceefb0edd03befd864d
<pre>
[Linux] TextureMapperPlatformLayerProxyDMABuf should support different colorspaces
<a href="https://bugs.webkit.org/show_bug.cgi?id=240633">https://bugs.webkit.org/show_bug.cgi?id=240633</a>

Patch by Žan Doberšek &lt;zdobersek@igalia.com &gt; on 2022-05-31
Reviewed by Miguel Gomez and Philippe Normand.

Add the DMABufColorSpace enumeration, covering different colorspaces we
currently can support between decoded GStreamer data and the TextureMapper
DMABuf integration.

DMABufObject gains a DMABufColorSpace member variable, initially of an invalid
value but that can be overridden by whoever is constructing the object.

In MediaPlayerPrivateGStreamer, the desired colorspace can be retrieved from the
GstVideoInfo colorimetry information and set on the DMABufObject.

The colorspace is now respected in TextureMapperPlatformLayerProxyDMABuf,
choosing between different YUV-to-RGB matrices that are to be used when sampling
from the set of plane-assigned textures. The default is the BT.601 conversion
matrix. For RGBA-like formats, no conversion is required, so an assert is placed
to ensure that the colorspace is either unspecified or specified as SRGB.

* Source/WebCore/platform/TextureMapper.cmake:
* Source/WebCore/platform/graphics/gbm/DMABufColorSpace.h: Added.
* Source/WebCore/platform/graphics/gbm/DMABufObject.h:
(WebCore::DMABufObject::encode const):
(WebCore::DMABufObject::encode):
(WebCore::DMABufObject::decode):
* Source/WebCore/platform/graphics/gbm/GBMBufferSwapchain.cpp:
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::colorSpaceForColorimetry):
(WebCore::MediaPlayerPrivateGStreamer::pushDMABufToCompositor):
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyDMABuf.cpp:
(WebCore::TextureMapperPlatformLayerProxyDMABuf::DMABufLayer::paintToTextureMapper):

Canonical link: <a href="https://commits.webkit.org/251143@main">https://commits.webkit.org/251143@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@295048">https://svn.webkit.org/repository/webkit/trunk@295048</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

philn commented on Source/WebCore/platform/graphics/gbm/DMABufColorSpace.h:
```diff
@@ -0,0 +1,42 @@
+
+namespace WebCore {
+
+enum class DMABufColorSpace : uint32_t {
```
Maybe reuse `PlatformVideoMatrixCoefficients` instead of redefining this enum?

philn replied:
Also `VideoTrackPrivateGStreamer::updateConfigurationFromCaps()` has some related code, perhaps some of it can be refactor to common facilities.

zdobersek replied:
They're not the same, in scope or intention. `PlatformVideoMatrixCoefficients` only describes coefficients, but same coefficients can be reused across different colorspaces. `DMABufColorSpace` is the exact colorspace used for the DMABuf object, which allows selecting the appropriate predefined conversion matrix, which can still be different for different colorspaces even if the same coefficients are used.

magomez reviewed (approved):
Changes look good to me. But maybe some assets can be reused, as Phil mentions.

zdobersek committed:
[GStreamer] DMABufVideoSink fails to list NV21 as a supported format
https://bugs.webkit.org/show_bug.cgi?id=241119

Patch by Žan Doberšek <zdobersek@igalia.com> on 2022-05-31
Reviewed by Miguel Gomez and Philippe Normand.

* Source/WebCore/platform/graphics/gstreamer/DMABufVideoSinkGStreamer.cpp:
NV12 is listed twice in the DMABufVideoSink's list of supported formats. This
is a typo, one of those was supposed to be NV21, a different-but-similar format
that works just fine.

Canonical link: https://commits.webkit.org/251141@main
git-svn-id: https://svn.webkit.org/repository/webkit/trunk@295046 268f45cc-cd09-0410-ab3c-d52691b4dbfc

aperezdc committed:
[WPE][GTK] Build broken with ENABLE_VIDEO=OFF after r292252
https://bugs.webkit.org/show_bug.cgi?id=241123

Reviewed by Philippe Normand.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp: Add missing ENABLE(VIDEO) guard.
* Source/WebKit/WebProcess/WebPage/WebPage.h: Ditto.

Canonical link: https://commits.webkit.org/251142@main
git-svn-id: https://svn.webkit.org/repository/webkit/trunk@295047 268f45cc-cd09-0410-ab3c-d52691b4dbfc

zdobersek committed:
[Linux] TextureMapperPlatformLayerProxyDMABuf should support different colorspaces
https://bugs.webkit.org/show_bug.cgi?id=240633

Patch by Žan Doberšek <zdobersek@igalia.com> on 2022-05-31
Reviewed by Miguel Gomez and Philippe Normand.

Add the DMABufColorSpace enumeration, covering different colorspaces we
currently can support between decoded GStreamer data and the TextureMapper
DMABuf integration.

DMABufObject gains a DMABufColorSpace member variable, initially of an invalid
value but that can be overridden by whoever is constructing the object.

In MediaPlayerPrivateGStreamer, the desired colorspace can be retrieved from the
GstVideoInfo colorimetry information and set on the DMABufObject.

The colorspace is now respected in TextureMapperPlatformLayerProxyDMABuf,
choosing between different YUV-to-RGB matrices that are to be used when sampling
from the set of plane-assigned textures. The default is the BT.601 conversion
matrix. For RGBA-like formats, no conversion is required, so an assert is placed
to ensure that the colorspace is either unspecified or specified as SRGB.

* Source/WebCore/platform/TextureMapper.cmake:
* Source/WebCore/platform/graphics/gbm/DMABufColorSpace.h: Added.
* Source/WebCore/platform/graphics/gbm/DMABufObject.h:
(WebCore::DMABufObject::encode const):
(WebCore::DMABufObject::encode):
(WebCore::DMABufObject::decode):
* Source/WebCore/platform/graphics/gbm/GBMBufferSwapchain.cpp:
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::colorSpaceForColorimetry):
(WebCore::MediaPlayerPrivateGStreamer::pushDMABufToCompositor):
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyDMABuf.cpp:
(WebCore::TextureMapperPlatformLayerProxyDMABuf::DMABufLayer::paintToTextureMapper):

Canonical link: https://commits.webkit.org/251143@main
git-svn-id: https://svn.webkit.org/repository/webkit/trunk@295048 268f45cc-cd09-0410-ab3c-d52691b4dbfc
